src/python/grpcio/grpc/experimental/aio/_interceptor.py

```diff
@@ -0,0 +1,270 @@
+    RPC has finished but again the execution is still held by the interceptors.
+
+    Once the RPC is finally executed, all methods are finally done against the
+    intercepted call, being at the same time the same call returned to the
+    interceptors.
+
+    For most of the methods, like `initial_metadata()` the caller does not need
+    to wait until the interceptors task is finished, once the RPC is done the
+    caller will have the freedom for accessing to the results.
+
+    For the `__await__` method is it is proxied to the intercepted call only when
+    the interceptor task is finished.
+    """
+
+    _loop: asyncio.AbstractEventLoop
+    _channel: cygrpc.AioChannel
+    _cancelled_before_rpc: bool
+    _intercepted_call: Optional[_base_call.UnaryUnaryCall]
+    _intercepted_call_created: asyncio.Event
+    _interceptors_task: asyncio.Task
+
+    def __init__(
+            self,
+            interceptors: Sequence[UnaryUnaryClientInterceptor],  # pylint: disable=R0913
+            request: RequestType,
+            timeout: Optional[float],
+            channel: cygrpc.AioChannel,
+            method: bytes,
+            request_serializer: SerializingFunction,
+            response_deserializer: DeserializingFunction) -> None:
+        self._channel = channel
+        self._loop = asyncio.get_event_loop()
+        self._intercepted_call = None
+        self._intercepted_call_created = asyncio.Event(loop=self._loop)
+        self._cancelled_before_rpc = False
+        self._invoke(interceptors, method, timeout, request, request_serializer,
+                     response_deserializer)
+
+    def __del__(self):
+        self.cancel()
+
+    def _invoke(self, interceptors: Sequence[UnaryUnaryClientInterceptor],
+                method: bytes, timeout: Optional[float], request: RequestType,
+                request_serializer: SerializingFunction,
+                response_deserializer: DeserializingFunction) -> None:
+        """Run the RPC call wrapped in interceptors"""
+
+        async def _run_interceptor(
+                interceptors: Iterator[UnaryUnaryClientInterceptor],
+                client_call_details: ClientCallDetails,
+                request: RequestType) -> _base_call.UnaryUnaryCall:
+            try:
+                interceptor = next(interceptors)
+            except StopIteration:
+                interceptor = None
+
+            if interceptor:
+                continuation = functools.partial(_run_interceptor, interceptors)
+                return await interceptor.intercept_unary_unary(
+                    continuation, client_call_details, request)
+            else:
+                self._intercepted_call = UnaryUnaryCall(
+                    request,
+                    _timeout_to_deadline(self._loop,
+                                         client_call_details.timeout),
+                    self._channel, client_call_details.method,
+                    request_serializer, response_deserializer)
```

Comment:
Alternative: We could plumb through the cancellation future here, and make Core generate a genuine CANCELLED status. In this way, we can reuse more code path.

Alternative 2: We could craft a function as final continuation that uses all the `client_call_details` variables and create the `UnaryUnaryCall`. In this way, the cancellation doesn't need to worry about before/after stuff. If cancellation happens before, the final continuation won't be executed; if cancellation happens after, the cancel should forward the `CancelledError` all the way down to Cython layer, hence cancelled normally.

Reply:
Correct me if I'm wrong, but what you are proposing is basically not managing the after and before, if the cancel happens and was done before the `UnaryUnaryCall` was created we just have to be sure that once the `UnaryUnaryCall` is created it realized that was already canceled. And if it happens while the RPC is still ongoing just follow the already existing path.

What would happen if the `cancel` is done when the RPC has already finished and the code execution is still held by the interceptor?

IMO I would manage cancellation only for the RPC, so doing the following:

- If cancelation happens before the RPC has invoked, just make sure that when is invoked it gets canceled immediately.
- if cancellation happens when the RPC is ongoing, cancel the RPC
- If cancellation happens when the RPC has finished, do nothing and consider the RPC finished and none cancellable

So I would not perturb the interceptor's execution during an RPC cancellation.

Reply:
Based on your design, alternative 2 requires a small change in the logic, and could simplify the logic here. By encapsulate the `UnaryUnaryCall` creation as the last continuation, we now have one task that representing the call the `self. _interceptors_task`.

It utilized the cancellation mechanism of `asyncio.Future` which generates an `CancelledError` exception in the deepest coroutine, then propagate upward to cancel all other coroutines.

#### Case 1: Before-RPC cancellation

In this case, the RPC is not initiated, once the cancel is called. The entire interception should stop. We could do that by calling `self._interceptors_task.cancel()` in side of `InterceptedUnaryUnaryCall`'s `cancel` method. In that way, the cancel will halt any ongoing interception, and let the `CancelledError` exception propagate.

#### Case 2: Ongoing-RPC cancellation

Once the RPC is running, it might pending on read or write. The `CancelledError` will be propagated to the Cython layer, and the RPC will be cancelled normally.

#### Case 3: After-RPC cancellation

After the RPC is finished, the final status is set, but the interception is not over. There will be a slightly strange behavior here, that the cancellation might appear success, but the final status is `StatusCode.OK`. The cancellation still effectively cancelled the second half of interception, but not setting the final status. To me, this is acceptable.

Reply:
What you are describing here is what is implemented right now if Im not missing something. Take a look at the following tests

- Case 1: Before-RPC cancellation https://github.com/grpc/grpc/pull/21455/files#diff-0894ffe42420c99925e97890b1f2685bR249
- Case 2: Ongoing-RPC cancellation (no test implemented) (already tested in other unit tests)
- Case 3: After-RPC cancellation https://github.com/grpc/grpc/pull/21455/files#diff-0894ffe42420c99925e97890b1f2685bR293

If you agree with that design, I would keep it. But TBH it has some weird behaviors, like the one that is described in the use case 3.

Reply:
@pfreixes Yes. The final behavior should be the same, I'm just suggesting a potential way to simplify the code. It should not be a blocker for this PR. Feel free to keep the implementation, and resolve other comments.